### PR TITLE
feat: add shutdown for datanode

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add `shutdown` method for datanode

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1158

### CODE REVIEW

File: src/table/src/engine.rs [hunk 1]
```diff
@@ -100,6 +100,11 @@ pub trait TableEngine: Send + Sync {
 
     /// Drops the given table. Return true if the table is dropped, or false if the table doesn't exist.
     async fn drop_table(&self, ctx: &EngineContext, request: DropTableRequest) -> Result<bool>;
+
+    /// Close the table.
+    ///
+    /// Returns number of deleted rows.
```
Comment (MichaelScofield) [inline src/table/src/engine.rs:106]:
```suggestion
```

File: src/table/src/table.rs [hunk 2]
```diff
@@ -93,6 +93,13 @@ pub trait Table: Send + Sync {
         }
         .fail()?
     }
+
+    /// Close the table.
+    ///
+    /// Returns number of deleted rows.
```
Comment (MichaelScofield) [inline src/table/src/table.rs:99]:
```suggestion
```

File: src/datanode/src/heartbeat.rs [hunk 3]
```diff
@@ -144,6 +144,19 @@ impl HeartbeatTask {
...
     }
+
+    pub async fn close(&self) -> Result<()> {
+        let running = self.running.clone();
+        if running
+            .compare_exchange(true, false, Ordering::AcqRel, Ordering::Acquire)
+            .is_err()
+        {
+            warn!("Call close heartbeat task multiple times");
+            return Ok(());
```
Comment (killme2008) [inline src/datanode/src/heartbeat.rs:155]: This `return` is unnecessary, it's returned at last.

File: src/table/src/table.rs [hunk 4]
```diff
@@ -93,6 +93,11 @@ pub trait Table: Send + Sync {
         }
         .fail()?
     }
+
+    /// Close the table.
+    async fn close(&self) -> Result<()> {
+        UnsupportedSnafu { operation: "CLOSE" }.fail()?
```
Comment (killme2008) [inline src/table/src/table.rs:99]: Prefer to return `Ok()` for default implementation.

File: src/mito/src/table.rs [hunk 5]
```diff
@@ -322,6 +322,19 @@ impl<R: Region> Table for MitoTable<R> {
...
+        if self.regions.is_empty() {
+            Ok(())
+        } else {
+            futures::future::try_join_all(self.regions.values().map(|region| region.close()))
+                .await
+                .map_err(BoxedError::new)
+                .context(table_error::TableOperationSnafu)?;
+
+            Ok(())
+        }
```
Comment (killme2008) [inline src/mito/src/table.rs:336]:
```suggestion
            futures::future::try_join_all(self.regions.values().map(|region| region.close()))
                .await
                .map_err(BoxedError::new)
                .context(table_error::TableOperationSnafu)
```

File: src/mito/src/engine.rs [hunk 6]
```diff
@@ -623,6 +627,28 @@ impl<S: StorageEngine> MitoEngineInner<S> {
...
+                res.push(table.close());
+            }
+
+            futures::future::try_join_all(res)
+                .await
+                .map_err(BoxedError::new)
+                .context(table_error::TableOperationSnafu)?;
+
+            Ok(())
+        }
```
Comment (killme2008) [inline src/mito/src/engine.rs:650]:
```suggestion
            futures::future::try_join_all(tables.values().map(|t| t.close()))
                .await
                .map_err(BoxedError::new)
                .context(table_error::TableOperationSnafu)
```

File: src/datanode/src/sql.rs [hunk 7]
```diff
@@ -139,6 +142,15 @@ impl SqlHandler {
...
         self.table_engine.clone()
     }
+
+    pub async fn close(&self) -> Result<()> {
+        self.table_engine
+            .close()
+            .await
+            .map_err(BoxedError::new)
+            .context(CloseTableEngineSnafu)?;
+        Ok(())
```
Comment (killme2008) [inline src/datanode/src/sql.rs:152]:
```suggestion
        self.table_engine
            .close()
            .await
            .map_err(BoxedError::new)
            .context(CloseTableEngineSnafu)
```

File: src/datanode/src/server.rs [hunk 8]
```diff
@@ -115,4 +117,16 @@ impl Services {
...
+    pub async fn shutdown(&self) -> Result<()> {
+        let mut res = vec![self.grpc_server.shutdown()];
+        if let Some(mysql_server) = &self.mysql_server {
+            res.push(mysql_server.shutdown());
+        }
+
+        futures::future::try_join_all(res)
+            .await
+            .context(ShutdownServerSnafu)?;
+        Ok(())
```
Comment (killme2008) [inline src/datanode/src/server.rs:130]:
```suggestion
        futures::future::try_join_all(res)
            .await
            .context(ShutdownServerSnafu)
```

File: src/datanode/src/instance.rs [hunk 9]
```diff
@@ -220,6 +221,24 @@ impl Instance {
...
+                .context(ShutdownInstanceSnafu)?;
+        }
+
+        self.sql_handler
+            .close()
+            .await
+            .map_err(BoxedError::new)
+            .context(ShutdownInstanceSnafu)?;
+
+        Ok(())
```
Comment (killme2008) [inline src/datanode/src/instance.rs:239]:
```suggestion
        self.sql_handler
            .close()
            .await
            .map_err(BoxedError::new)
            .context(ShutdownInstanceSnafu)
```

File: src/datanode/src/datanode.rs [hunk 10]
```diff
@@ -240,6 +240,20 @@ impl Datanode {
     pub fn get_instance(&self) -> InstanceRef {
         self.instance.clone()
     }
+
+    pub async fn shutdown_instance(&self) -> Result<()> {
```
Comment (killme2008) [inline src/datanode/src/datanode.rs:244]: Not necessary to be public.
```suggestion
   async fn shutdown_instance(&self) -> Result<()> {
```

File: src/datanode/src/datanode.rs [hunk 11]
```diff
@@ -240,6 +240,20 @@ impl Datanode {
...
+        self.instance.shutdown().await
+    }
+
+    pub async fn shutdown_services(&self) -> Result<()> {
```
Comment (killme2008) [inline src/datanode/src/datanode.rs:248]:
```suggestion
   async fn shutdown_services(&self) -> Result<()> {
```

File: src/datanode/src/instance.rs (hunk 9 above)
Comment (WenyXu) [reply src/datanode/src/instance.rs:239]: > Suggested change
>         self.sql_handler
>             .close()
>             .await
>             .map_err(BoxedError::new)
>             .context(ShutdownInstanceSnafu)?;
>         Ok(())
>         self.sql_handler
>             .close()
>             .await
>             .map_err(BoxedError::new)
>             .context(ShutdownInstanceSnafu)

It seems like leading to a mismatched types error:
```
mismatched types
expected enum `std::result::Result<(), _>`
   found enum `std::result::Result<Vec<()>, _>`
```

Comment (killme2008) [reply src/datanode/src/instance.rs:239]: Goit

File: src/datanode/src/server.rs (hunk 8 above)
Comment (WenyXu) [reply src/datanode/src/server.rs:130]: > Suggested change
>         futures::future::try_join_all(res)
>             .await
>             .context(ShutdownServerSnafu)?;
>         Ok(())
>         futures::future::try_join_all(res)
>             .await
>             .context(ShutdownServerSnafu)

It seems like leading to a mismatched types error:
```
mismatched types
expected enum `std::result::Result<(), _>`
   found enum `std::result::Result<Vec<()>, _>`
```

File: src/datanode/src/instance.rs (hunk 9 above)
Comment (WenyXu) [reply src/datanode/src/instance.rs:239]: > > Suggested change
> > self.sql_handler
> > .close()
> > .await
> > .map_err(BoxedError::new)
> > .context(ShutdownInstanceSnafu)?;
> > Ok(())
> > self.sql_handler
> > .close()
> > .await
> > .map_err(BoxedError::new)
> > .context(ShutdownInstanceSnafu)
> 
> It seems like leading to a mismatched types error:
> 
> ```
> mismatched types
> expected enum `std::result::Result<(), _>`
>    found enum `std::result::Result<Vec<()>, _>`
> ```

My bad, not this one

File: src/datanode/src/heartbeat.rs (hunk 3 above)
Comment (WenyXu) [reply src/datanode/src/heartbeat.rs:155]: > This `return` is unnecessary, it's returned at last.

Also leading to mismatched types error: 
```
error[E0308]: mismatched types
   --> src/datanode/src/heartbeat.rs:155:13
    |
150 | /         if running
151 | |             .compare_exchange(true, false, Ordering::AcqRel, Ordering::Acquire)
152 | |             .is_err()
153 | |         {
154 | |             warn!("Call close heartbeat task multiple times");
155 | |             Ok(())
    | |             ^^^^^^ expected `()`, found `Result<(), _>`
156 | |         }
    | |_________- expected this to be `()`
    |
    = note: expected unit type `()`
                    found enum `std::result::Result<(), _>`
help: consider using a semicolon here
    |
155 |             Ok(());
    |                   +
help: consider using a semicolon here
    |
156 |         };
    |          +
help: you might have meant to return this value
    |
155 |             return Ok(());
    |             ++++++       +

For more information about this error, try `rustc --explain E0308`.
```

Or change it to the following?
```rust
    pub async fn close(&self) -> Result<()> {
        let running = self.running.clone();
        if running
            .compare_exchange(true, false, Ordering::AcqRel, Ordering::Acquire)
            .is_err()
        {
            warn!("Call close heartbeat task multiple times");
            Ok(())
        } else {
            Ok(())
        }
    }
```

Comment (WenyXu) [reply src/datanode/src/heartbeat.rs:155]: > What about:
> 
> ```rust
> let running = self.running.clone();
>         if running
>             .compare_exchange(true, false, Ordering::AcqRel, Ordering::Acquire)
>             .is_err()
>         {
>             warn!("Call close heartbeat task multiple times");
>         } 
>        Ok(())
> ```
👌

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**